### ISSUE TITLE
clippy::swap_with_temporary false positive

### ISSUE THREAD

**Issue by hniksic** [rust-lang/rust-clippy#15472]
### Summary

`clippy::swap_with_temporary` misfires on code where the suggested assignment wouldn't work.

### Reproducer

I have code like this:

```rust
    std::mem::swap(
        &mut result.sets,
        &mut DYNAMIC_SETS.get_or_default().lock().unwrap(),
    );
```

clippy reports the following:

```none
warning: swapping with a temporary value is inefficient
   --> stats/src/counters.rs:389:5
    |
389 | /     std::mem::swap(
390 | |         &mut result.sets,
391 | |         &mut DYNAMIC_SETS.get_or_default().lock().unwrap(),
392 | |     );
    | |_____^ help: use assignment instead: `result.sets = DYNAMIC_SETS.get_or_default().lock().unwrap()`
    |
note: this expression returns a temporary value
   --> stats/src/counters.rs:391:14
    |
391 |         &mut DYNAMIC_SETS.get_or_default().lock().unwrap(),
    |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#swap_with_temporary
    = note: `#[warn(clippy::swap_with_temporary)]` on by default
```

but applying that suggestions leads to a compile error:

```none
error[E0308]: mismatched types
   --> stats/src/counters.rs:389:19
    |
389 |     result.sets = DYNAMIC_SETS.get_or_default().lock().unwrap();
    |     -----------   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `HashMap<Cow<'_, str>, HashMap<..., u64>>`, found `MutexGuard<'_, HashMap<..., ...>>`
    |     |
    |     expected due to the type of this binding
    |
    = note: expected struct `HashMap<_, _>`
               found struct `MutexGuard<'_, HashMap<_, _>>`

For more information about this error, try `rustc --explain E0308`.
```

Even if there were no error, swapping and assigning aren't really the same thing, so I guess I don't understand this lint. The [documentation](https://rust-lang.github.io/rust-clippy/stable/index.html#swap_with_temporary) uses contrived examples where the whole value is dropped right after the swap, but in reality "temporary" expressions can and do refer to places that will be observed later, as in the case above.

### Version

```text
rustc -vV
rustc 1.88.0 (6b00bc388 2025-06-23) (built from a source tarball)
binary: rustc
commit-hash: 6b00bc3880198600130e1cf62b8f8a93494488cc
commit-date: 2025-06-23
host: x86_64-unknown-linux-gnu
release: 1.88.0
LLVM version: 20.1.6
```

**Comment by hniksic** [comment]
Here is a standalone reproducer:

```rust
#![allow(clippy::type_complexity)]

use std::borrow::Cow;
use std::collections::HashMap;
use std::sync::{Mutex, OnceLock};

fn main() {
    static DYNAMIC_SETS: OnceLock<
        Mutex<HashMap<Cow<'static, str>, HashMap<Cow<'static, str>, u64>>>,
    > = OnceLock::new();
    let mut sets: HashMap<Cow<'static, str>, HashMap<Cow<'static, str>, u64>> = Default::default();
    std::mem::swap(
        &mut sets,
        &mut DYNAMIC_SETS.get_or_init(Default::default).lock().unwrap(),
    );
}
```

Output:

```none
cargo clippy
    Checking playground v0.1.0 (/home/hniksic/work/playground)
warning: swapping with a temporary value is inefficient
  --> src/main.rs:12:5
   |
12 | /     std::mem::swap(
13 | |         &mut sets,
14 | |         &mut DYNAMIC_SETS.get_or_init(Default::default).lock().unwrap(),
15 | |     );
   | |_____^ help: use assignment instead: `sets = DYNAMIC_SETS.get_or_init(Default::default).lock().unwrap()`
   |
note: this expression returns a temporary value
  --> src/main.rs:14:14
   |
14 |         &mut DYNAMIC_SETS.get_or_init(Default::default).lock().unwrap(),
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#swap_with_temporary
   = note: `#[warn(clippy::swap_with_temporary)]` on by default

warning: `playground` (bin "playground") generated 1 warning (run `cargo clippy --fix --bin "playground"` to apply 1 suggestion)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.12s
```

**Comment by hniksic** [comment]
I've now tried the standalone example with the latest stable Rust (1.89), and the false positive is no longer reported. Closing the issue.
